streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -764,6 +764,53 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
       } yield pull
     }
 
+  private final def bufferSignal[E1 >: E, A1 >: A](
+    queue: Queue[(Take[E1, A1], Promise[Nothing, Unit])]
+  ): ZManaged[R, Nothing, Pull[R, E1, A1]] =
+    for {
+      as    <- self.process
+      start <- Promise.make[Nothing, Unit].toManaged_
+      _     <- start.succeed(()).toManaged_
+      ref   <- Ref.make(start).toManaged_
+      done  <- Ref.make(false).toManaged_
+      upstream = {
+        def offer(take: Take[E1, A1]): UIO[Unit] = take match {
+          case Take.Value(_) =>
+            for {
+              p     <- Promise.make[Nothing, Unit]
+              added <- queue.offer((take, p))
+              _     <- ref.set(p).when(added)
+            } yield ()
+
+          case _ =>
+            for {
+              latch <- ref.get
+              _     <- latch.await
+              p     <- Promise.make[Nothing, Unit]
+              _     <- queue.offer((take, p))
+              _     <- ref.set(p)
```

Comment:
what if offer returns false in this case?

Reply:
Good question. It will not happen, because this code is blocked until the queue is drained (it is always waiting on the last enqueued promise, and the taker always unblocks promises as it encounters them), so there will always be space in the queue for this offer, no matter its type.